src/FsCodec/StreamName.fs

```diff
@@ -15,116 +15,88 @@ and [<Measure>] streamName
-            if rawCategory.IndexOf '-' <> -1 then invalidArg "rawCategory" "may not contain embedded '-' symbols"
+        let [<Literal>] Separator = '_' // separates {category}-{streamId}
+        let [<Literal>] SeparatorStr = "_"
+        let internal separator = [| Separator |]
 
-        /// Throws if a candidate id element includes a '_', is null, or is empty
-        let inline validateElement (rawElement : string) =
-            if rawElement |> System.String.IsNullOrEmpty then invalidArg "rawElement" "may not contain null or empty components"
-            if rawElement.IndexOf '_' <> -1 then invalidArg "rawElement" "may not contain embedded '_' symbols"
+        /// Throws if a candidate category includes a '-', is null, or is empty
+        let inline validate (raw: string) =
+            if raw |> System.String.IsNullOrEmpty then invalidArg "raw" "Category must not be null or empty"
+            if raw.IndexOf Separator <> -1 then invalidArg "raw" "Category must not contain embedded '-' symbols"
 
-        /// Low level helper used to gate ingestion from a canonical form, guarding against malformed streamNames
-        let inline ofCategoryAndStreamId struct (category : string, streamId : string) : string =
-            validateCategory category
-            System.String.Concat(category, "-", streamId)
+        /// Extracts the category portion of the StreamName
+        let ofStreamName (x: StreamName) =
+            let raw = toString x
+            raw.Substring(0, raw.IndexOf Separator)
 
-        /// Generates a StreamId from name elements; elements are separated from each other by '_'
-        let createStreamId (elements : string seq) : string =
-            for x in elements do validateElement x
-            System.String.Join("_", elements)
+    module Internal =
 
-    (* Creators: Building from constituent parts
-       Guards against malformed category, streamId and/or streamId elements with exceptions *)
+        /// Create a StreamName, trusting the input to be well-formed
+        let trust (raw: string): StreamName = UMX.tag raw
 
-    /// Recommended way to specify a stream identifier; a category identifier and an streamId representing the aggregate's identity
-    /// category is separated from id by `-`
-    let create (category : string) (streamId : string) : StreamName =
-        Internal.ofCategoryAndStreamId (category, streamId) |> UMX.tag
+        /// <summary>Attempts to split a Stream Name in the form <c>{category}-{streamId}</c> into its two elements.
+        /// The <c>{streamId}</c> segment is permitted to include embedded '-' (dash) characters
+        /// Returns <c>None</c> if it does not adhere to that form.</summary>
+        let tryParse (raw: string): struct (string * StreamId) voption =
+            match raw.Split(Category.separator, 2) with
+            | [| cat; id |] -> ValueSome struct (cat, StreamId.Elements.trust id)
```

Comment:
id is not guaranteed to contain no `_` at this point

Reply:
```
> "a-b_c".Split("-", 5);;
val it: string array = [|"a"; "b_c"|]
```

Reply:
Oh I see where you're going, this is expected. StreamName is just dealing with the `-` split, you have another layer `StreamId` dealing with the rest :+1:

Reply:
we guard against embedded `-` chars in the cat on creation
but the bottom line is anything up to the first `-` is the cat. The rest is the streamid
And we are returning it as a streamid, which is allowed to contain `_`s (but normal composition ensures you wont be using embedded `_` within them
Maybe I'm not getting your point ;)

Reply:
ha, sorry, missed that comment, bad wifi...

Do keep looking though - I'll validate this with Equinox and Propulsion but the next steps are:
- see in the context of an app
- see what can be learned from levelling with the JS
- get the tests more complete
there's def potential to improve naming and/or simplify here